runtime/fiber.c

```diff
@@ -218,18 +225,21 @@ void caml_change_max_stack_size (uintnat new_max_size)
 void caml_scan_stack(scanning_action f, void* fdata, struct stack_info* stack)
 {
   value *low, *high, *sp;
-  Assert(stack->magic == 42);
-
-  f(fdata, Stack_handle_value(stack), &Stack_handle_value(stack));
-  f(fdata, Stack_handle_exception(stack), &Stack_handle_exception(stack));
-  f(fdata, Stack_handle_effect(stack), &Stack_handle_effect(stack));
-  if (Stack_parent(stack))
-    caml_scan_stack(f, fdata, Stack_parent(stack));
-
-  high = Stack_high(stack);
-  low = stack->sp;
-  for (sp = low; sp < high; sp++) {
-    f(fdata, *sp, sp);
+
+  while (stack != NULL) {
+    Assert(stack->magic == 42);
+
+    high = Stack_high(stack);
```

Comment:
There's a slight inversion of operations here regarding the order in which nodes are visited, however Tom confirmed me this is fine.